airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/async_retriever.py

```diff
@@ -0,0 +1,113 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+
+from dataclasses import InitVar, dataclass, field
+from typing import Any, Callable, Iterable, List, Mapping, Optional, Union
+
+from airbyte_cdk.models import FailureType
+from airbyte_cdk.sources.declarative.async_job.job_orchestrator import AsyncJobOrchestrator, AsyncPartition
+from airbyte_cdk.sources.declarative.extractors.record_selector import RecordSelector
+from airbyte_cdk.sources.declarative.interpolation import InterpolatedString
+from airbyte_cdk.sources.declarative.partition_routers import SinglePartitionRouter
+from airbyte_cdk.sources.declarative.retrievers import Retriever
+from airbyte_cdk.sources.declarative.stream_slicers import StreamSlicer
+from airbyte_cdk.sources.streams.core import StreamData
+from airbyte_cdk.sources.types import Config, StreamSlice, StreamState
+from airbyte_cdk.utils.traced_exception import AirbyteTracedException
+
+
+@dataclass
+class AsyncRetriever(Retriever):
+    config: Config
+    parameters: InitVar[Mapping[str, Any]]
+    job_orchestrator_factory: Callable[[Iterable[StreamSlice]], AsyncJobOrchestrator]
+    record_selector: RecordSelector
+    stream_slicer: StreamSlicer = field(default_factory=lambda: SinglePartitionRouter(parameters={}))
+
+    def __post_init__(self, parameters: Mapping[str, Any]) -> None:
+        self._job_orchestrator_factory = self.job_orchestrator_factory
+        self.__job_orchestrator: Optional[AsyncJobOrchestrator] = None
+        self._parameters = parameters
+
+    @property
+    def state(self) -> StreamState:
+        """
+        As a first iteration for sendgrid, there is no state to be managed
+        """
+        return {}
```

Comment:
Q: is this a good time to pull the state management out of the retriever into the DeclarativeStream?

Reply:
Given the deadlines we have for async project, I would say this is out of scope but we can have the discussion

Reply:
making the consequence of this decision explicit: async retriever only supports full refresh streams. is that right?